Batteries/Data/Fin/Lemmas.lean

```diff
@@ -13,3 +13,143 @@ attribute [norm_cast] val_last
 /-! ### clamp -/
 
 @[simp] theorem coe_clamp (n m : Nat) : (clamp n m : Nat) = min n m := rfl
+
+/-! ### findSome? -/
+
+@[simp] theorem findSome?_zero {f : Fin 0 → Option α} : findSome? f = none := rfl
+
+@[simp] theorem findSome?_one {f : Fin 1 → Option α} : findSome? f = f 0 := rfl
+
+theorem findSome?_succ {f : Fin (n+1) → Option α} :
+    findSome? f = (f 0 <|> findSome? fun i => f i.succ) := by
+  simp only [findSome?, foldl_succ, Option.none_orElse, Function.comp_apply]
+  cases f 0
+  · rw [Option.none_orElse]
+  · rw [Option.some_orElse]
+    induction n with
+    | zero => rfl
+    | succ n ih => rw [foldl_succ, Option.some_orElse, ih (f := fun i => f i.succ)]
+
+@[simp] theorem findSome?_succ_of_some {f : Fin (n+1) → Option α} (h : f 0 = some x) :
+    findSome? f = some x := by simp [findSome?_succ, h]
+
+@[simp] theorem findSome?_succ_of_isSome {f : Fin (n+1) → Option α} (h : (f 0).isSome) :
+    findSome? f = f 0 := by cases _h : f 0 <;> simp_all
+
+@[simp] theorem findSome?_succ_of_none {f : Fin (n+1) → Option α} (h : f 0 = none) :
+    findSome? f = findSome? fun i => f i.succ := by simp [findSome?_succ, h]
+
+@[simp] theorem findSome?_succ_of_isNone {f : Fin (n+1) → Option α} (h : (f 0).isNone) :
+    findSome? f = findSome? fun i => f i.succ := by simp_all
+
+theorem exists_of_findSome?_eq_some {f : Fin n → Option α} (h : findSome? f = some x) :
+    ∃ i, f i = some x := by
+  induction n with
+  | zero => rw [findSome?_zero] at h; contradiction
+  | succ n ih =>
+    rw [findSome?_succ] at h
+    match heq : f 0 with
+    | some x =>
+      rw [heq, Option.some_orElse] at h
+      exists 0
+      rw [heq, h]
+    | none =>
+      rw [heq, Option.none_orElse] at h
+      match ih h with | ⟨i, _⟩ => exists i.succ
+
+theorem eq_none_of_findSome?_eq_none {f : Fin n → Option α} (h : findSome? f = none) (i) :
+    f i = none := by
+  induction n with
+  | zero => cases i; contradiction
+  | succ n ih =>
+    rw [findSome?_succ] at h
+    match heq : f 0 with
+    | some x =>
+      rw [heq, Option.some_orElse] at h
+      contradiction
+    | none =>
+      rw [heq, Option.none_orElse] at h
+      cases i using Fin.cases with
+      | zero => exact heq
+      | succ i => exact ih h i
+
+@[simp] theorem findSome?_isSome_iff {f : Fin n → Option α} :
+    (findSome? f).isSome ↔ ∃ i, (f i).isSome := by
+  simp only [Option.isSome_iff_exists]
+  constructor
+  · intro ⟨x, hx⟩
+    match exists_of_findSome?_eq_some hx with
+    | ⟨i, hi⟩ => exists i, x
+  · intro ⟨i, x, hix⟩
+    match h : findSome? f with
+    | some x => exists x
+    | none => rw [eq_none_of_findSome?_eq_none h i] at hix; contradiction
+
+@[simp] theorem findSome?_eq_none_iff {f : Fin n → Option α} :
```

Comment:
What about the missing `findSome?_eq_some_iff`, which also says that `f` is none for earlier values?

Reply:
You could hopefully prove this via `findSome?_eq_findSome?_finRange` and `List.findSome?_eq_some_iff`.

Reply:
Just checking: this is about `findSome?_isSome_iff` and not about `findSome?_eq_none_iff`. The code selection is a bit off.

Anyway, my current proof of this sucks. I'm busy this week so I'm adding a `proof_wanted` to temporarily deal with this. I'll come back to it if nobody fixes this before I do.